demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_csr_demo.c

```diff
@@ -0,0 +1,1271 @@
+ * @param[in,out] certificateIdLength The length of #certificateId. The written
+ * length is output here.
+ * @param[in] ownershipToken The buffer to which to write the certificate
+ * ownership token.
+ * @param[in,out] ownershipTokenLength The length of #ownershipToken. The written
+ * length is output here.
+ */
+static bool parseCsrResponse( const char * response,
+                              size_t length,
+                              char * certificate,
+                              size_t * certificateLength,
+                              char * certificateId,
+                              size_t * certificateIdLength,
+                              char * ownershipToken,
+                              size_t * ownershipTokenLength );
+
+/**
+ * @brief Extracts the Thing name from a RegisterThing accepted response.
+ *
+ * @param[in] response The response document.
+ * @param[in] length Length of #response.
+ * @param[in] thingNameBuffer The buffer to which to write the Thing name.
+ * @param[in,out] thingNameBufferLength The length of #thingNameBuffer. The written
+ * length is output here.
+ */
+static bool parseRegisterThingResponse( const char * response,
+                                        size_t length,
+                                        char * thingNameBuffer,
+                                        size_t * thingNameBufferLength );
+
+/**
+ * @brief Run the MQTT process loop to get a response.
+ */
+static bool waitForResponse( void );
+
+/**
+ * @brief Subscribe to the CreateCertificateFromCsr accepted and rejected topics.
+ */
+static bool subscribeToCsrResponseTopics( void );
+
+/**
+ * @brief Unsubscribe from the CreateCertificateFromCsr accepted and rejected topics.
+ */
+static bool unsubscribeFromCsrResponseTopics( void );
+
+/**
+ * @brief Subscribe to the RegisterThing accepted and rejected topics.
+ */
+static bool subscribeToRegisterThingResponseTopics( void );
+
+/**
+ * @brief Unsubscribe from the RegisterThing accepted and rejected topics.
+ */
+static bool unsubscribeFromRegisterThingResponseTopics( void );
+
+/**
+ * @brief Save the certificate to the path specified by #PROVISIONING_CERT_PATH.
+ */
+static bool saveCertificate( const char * certificate,
+                             size_t certificateLength );
+/*-----------------------------------------------------------*/
+
+static void provisioningPublishCallback( MQTTPublishInfo_t * pPublishInfo,
+                                         uint16_t packetIdentifier )
+{
+    FleetProvisioningStatus_t status;
+    FleetProvisioningTopic_t api;
+
+    /* Silence compiler warnings about unused variables. */
+    ( void ) packetIdentifier;
+
+    status = FleetProvisioning_MatchTopic( pPublishInfo->pTopicName,
+                                           pPublishInfo->topicNameLength, &api );
+
+    if( status != FleetProvisioningSuccess )
+    {
+        LogError( ( "Unexpected publish message received. Topic: %.*s.",
```

Comment:
A warning level of log message suffices as it does not affect the demo from proceeding further. 
```suggestion
        LogWarning( ( "Unexpected publish message received. Topic: %.*s.",
```